nemo_asr/run_eval.py

```diff
@@ -8,131 +8,142 @@
+    
+    asr_model.to(compute_dtype)
+    asr_model.eval()
 
     dataset = data_utils.load_data(args)
 
+    def download_audio_files(batch):
+
+        # download audio files and write the paths, transcriptions and durations to a manifest file
+        audio_paths = []
+        durations = []
+
+        for id, sample in zip(batch["id"], batch["audio"]):
+            audio_path = os.path.join(CACHE_DIR, f"{id}.wav")
+            if not os.path.exists(audio_path):
+                os.makedirs(os.path.dirname(audio_path), exist_ok=True)
+                soundfile.write(audio_path, np.float32(sample["array"]), 16_000)
+            audio_paths.append(audio_path)
+            durations.append(len(sample["array"]) / 16_000)
+
+        
+        batch["references"] = batch["norm_text"]
+        batch["audio_filepaths"] = audio_paths
+        batch["durations"] = durations
+
+        return batch
+
+
     if args.max_eval_samples is not None and args.max_eval_samples > 0:
         print(f"Subsampling dataset to first {args.max_eval_samples} samples !")
         dataset = dataset.take(args.max_eval_samples)
 
     dataset = data_utils.prepare_data(dataset)
-
-    predictions = []
-    references = []
-
-    # run streamed inference
-    cache_prefix = (f"{args.model_id.replace('/', '-')}-{args.dataset_path.replace('/', '')}-"
-                    f"{args.dataset.replace('/', '-')}-{args.split}")
-    results = buffer_audio_and_transcribe(asr_model, dataset, args.batch_size, args.pnc, cache_prefix, verbose=True)
-    for sample in results:
-        predictions.append(data_utils.normalizer(sample["pred_text"]))
-        references.append(sample["reference"])
-
-    # Write manifest results
+    if asr_model.cfg.decoding.strategy != "beam":
+        asr_model.cfg.decoding.strategy = "greedy_batch"
+        asr_model.change_decoding_strategy(asr_model.cfg.decoding)
+    
+    # prepraing the offline dataset
+    dataset = dataset.map(download_audio_files, batch_size=args.batch_size, batched=True, remove_columns=["audio"])
+
+    # Write manifest from daraset batch using json and keys audio_filepath, duration, text
+
+    all_data = {
+        "audio_filepaths": [],
+        "durations": [],
+        "references": [],
+    }
+
+    data_itr = iter(dataset)
+    for data in tqdm(data_itr, desc="Downloading Samples"):
+        # import ipdb; ipdb.set_trace()
+        for key in all_data:
+            all_data[key].append(data[key])
+    
+    # Sort audio_filepaths and references based on durations values
+    sorted_indices = sorted(range(len(all_data["durations"])), key=lambda k: all_data["durations"][k], reverse=True)
+    all_data["audio_filepaths"] = [all_data["audio_filepaths"][i] for i in sorted_indices]
+    all_data["references"] = [all_data["references"][i] for i in sorted_indices]
+    all_data["durations"] = [all_data["durations"][i] for i in sorted_indices]
+    
+    
+    total_time = 0
+    for _ in range(2): # warmup once and calculate rtf 
```

Comment:
Do we need to warm-up over the entire dataset? Would a subset of say 10 batches (i.e. 10 warm-up steps) not suffice?

Reply:
good point, updated to warming upto 4 batches